Fixed array length

svenk177 opened the pull request:
This change ports the pull request #3987 by daksenik and idoroshev to the most recent commit and adds support for generation of JSON schema, Java, C# and Python.

vglavnyy commented:
Could you add an example of fbs and json as part of `monster_extra`?
C++, C#, Python have the support of `extra`.

1. Why not use an attribute for array size specification?
For example:
```
struct A{
 B:[float] (fixed_length:16);
}
```

2. Is Array type really needed?
The Array looks like Vector, therefore can reuse 90% of Vector code.
Can we embed "uoffset_t length_" field into the struct to emulate Vector?
If one needs Array the four bytes isn't big overhead.

svenk177 commented:
> Could you add an example of fbs and json as part of `monster_extra`?
> C++, C#, Python have the support of `extra`.

An example to monster_extra was added. It seems the JSON generator crashes on parsing the floating point extensions, therefore a JSON schema is generated from array_test.fbs for now.
 
> 1. Why not use an attribute for array size specification?
>    For example:
> 
> ```
> struct A{
>  B:[float] (fixed_length:16);
> }
> ```

This syntax is derived from a proposed feature for [flatcc](https://github.com/dvidelabs/flatcc/blob/master/doc/binary-format.md#fixed-length-vectors).
NOTE: This patch currently only supports primitive types in structs!

> 2. Is Array type really needed?
>    The Array looks like Vector, therefore can reuse 90% of Vector code.
>    Can we embed "uoffset_t length_" field into the struct to emulate Vector?
>    If one needs Array the four bytes isn't big overhead.

Again this was derived from flatcc (see above). The expression 
```
array:[int:2];
```
shall be equivalent to 
```
int1:int;
int2:int;
```
in its binary representation.

aardappel commented on tests/MyGame/MonsterExtra.cs:
```diff
@@ -29,31 +29,16 @@ public struct MonsterExtra : IFlatbufferObject
   public bool MutateTestdNinf(double testd_ninf) { int o = __p.__offset(14); if (o != 0) { __p.bb.PutDouble(o + __p.bb_pos, testd_ninf); return true; } else { return false; } }
+  public ArrayStruct? Testa { get { int o = __p.__offset(16); return o != 0 ? (ArrayStruct?)(new ArrayStruct()).__assign(o + __p.bb_pos, __p.bb) : null; } }
 
-  public static Offset<MonsterExtra> CreateMonsterExtra(FlatBufferBuilder builder,
```
Is this constructor being deleted because we don't generate them for structs that have arrays in them? If so that is fine I guess.

svenk177 replied:
This constructor seems to be deleted as soon as a struct is added to a table. This is not related to the patch set.

aardappel commented on tests/MyGame/MonsterExtra.py:
```diff
@@ -60,11 +60,23 @@ def TestdNinf(self):
+        o = flatbuffers.number_types.UOffsetTFlags.py_type(self._tab.Offset(16))
+        if o != 0:
+            x = o + self._tab.Pos
+            from .ArrayStruct import ArrayStruct
```
why is this import in the middle of the code? is that a normal thing to do in Python? :)

svenk177 replied:
Again this is seems to be generated when a struct is added to a table. It was not introduced with this patch set :)

aardappel commented on tests/arrays_test.fbs:
```diff
@@ -0,0 +1,22 @@
+namespace MyGame.Example;
+
+struct ArrayStruct{
+	a:float;
```
consistently indent 2 spaces.

aardappel commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,225 @@
+        c_(flatbuffers::EndianScalar(_c)),
+        padding0__(0),
+        padding1__(0) {
+    memcpy(b_, _b, 60);
```
This will not work on big endian machines, you'll need a loop with `EndianScalar` I'm afraid. Maybe make it a helper templated function in `flatbuffers.h`.

aardappel commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,225 @@
+  void mutate_a(float _a) {
+    flatbuffers::WriteScalar(&a_, _a);
+  }
+  const int32_t *b() const {
```
This is also endian unsafe. Make a class like `Vector` that does all access.

svenk177 commented on src/idl_parser.cpp:
```diff
@@ -1229,6 +1262,59 @@ CheckedError Parser::ParseVector(const Type &type, uoffset_t *ovalue,
+    if (token_ != ',')
+    {
+        ECHECK(ParseSingleValue(nullptr, val, false));
+        stack.push_back(val);
```
ParseSingleValue has support for trigonometric functions, which seem pretty useful.

aardappel replied:
I am not saying to not use `ParseSingleValue`, I am saying not to allocate a whole stack of `Value`. You can instead use a single `Value` and instead store a buffer of the actual binary values.

svenk177 commented on tests/MyGame/ArrayStruct.java:
```diff
@@ -0,0 +1,24 @@
+  public void __init(int _i, ByteBuffer _bb) { bb_pos = _i; bb = _bb; }
+  public ArrayStruct __assign(int _i, ByteBuffer _bb) { __init(_i, _bb); return this; }
+
+  public int [] array() { int [] b = new int[2]; for (int i = 0; i < 2; i++) { b[i] = bb.getInt(bb_pos + 0 + i * 4); } return b; }
```
I'm not quite sure here. Java and C# always require dynamic allocation even for wrapper classes.

aardappel replied:
Yes, but those can be reused. Array accesses are potentially performance sensitive, so lets follow the existing vector API design.

svenk177 commented:
Thank you for your kind response. I introduced most of the requested changes and I left some comments on the remaining ones.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -393,6 +393,84 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
+  // Get a mutable pointer to tables/strings inside this array.
+  mutable_return_type GetMutableObject(uoffset_t i) const {
+    FLATBUFFERS_ASSERT(i < size());
+    return const_cast<mutable_return_type>(IndirectHelper<T>::Read(Data(), i));
```
same here

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -393,6 +393,84 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
+
+  return_type Get(uoffset_t i) const {
+    FLATBUFFERS_ASSERT(i < size());
+    return IndirectHelper<T>::Read(Data(), i);
```
Weird, github somehow deleted my comment that "same here" refers to.
In short: lets not use `IndirectHelper` since unlike `Vector` we don't have complex types to dereference.

mikkelfj commented on src/idl_parser.cpp:
```diff
@@ -590,12 +590,33 @@ CheckedError Parser::ParseType(Type &type) {
+              "length of fixed-length array must be an integer value");
+      }
+      int64_t fixed_length = StringToInt(attribute_.c_str());
+      if (fixed_length < 1 || fixed_length > 0x7fff) {
```
https://stackoverflow.com/a/7724909 says 2^16-1 is a C limit on object size, so it would be reasonable to limit the size to 16 bits, though I'm not sure about a 2^15-1 length limit. I'd rather place a limit on struct sizes of 2^16-1 and forego array length limits.

mikkelfj commented:
Does the syntax permit hex size?
```
foo[x:0x20];
```
flatcc generally accepts hex values in the schema parser where integers are permitted.

mikkelfj commented:
On alignment: I think alignment should be to the element size of the fixed array not the array size as such. The statement of equivalence to repeated scalar fields implies that.

On lenght limit:

flatcc already imposes the following struct limit by default.

While this limit is derived fro flatbuffers voffset_t size, it incidentally happens to the limit set by the C standard for portable struct sizes.

I don't think the 0x7fff length limit is a good idea. It would prevent 32K lookup tables.

config/config.h 
```
/*
 * There are no hard limits on structs, but
 * they cannot be effectively handled by tables
 * if larger than 64K and 64K-1 allows us to store
 * the size in short where needed. Internally
 * we can handle 64 bits.
 */
#ifndef FLATCC_STRUCT_MAX_SIZE
#define FLATCC_STRUCT_MAX_SIZE ((1 << 16) - 1)
#endif
```

mikkelfj commented:
We probably also want to permit fixed size arrays of structs as they can be used as typedefs.

mikkelfj commented:
I'm not sure if flatc has key attributes on struct fields (flatcc does), but they probably should not be allowed on fixed size arrays.

vglavnyy commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -393,6 +393,84 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
 }
 
+// This is used as a helper type for accessing arrays.
+template<typename T, int length> class Array {
```
The type of length should be matched with:
- uoffset_t size() const { return length; }
- short Type::fixed_length

svenk177 replied:
Changed to `uint16_t`.

vglavnyy commented on include/flatbuffers/idl.h:
```diff
@@ -167,6 +170,7 @@ struct Type {
   StructDef *struct_def;  // only set if t or element == BASE_TYPE_STRUCT
   EnumDef *enum_def;      // set if t == BASE_TYPE_UNION / BASE_TYPE_UTYPE,
                           // or for an integral type derived from an enum.
+  short fixed_length;     // only set if t == BASE_TYPE_ARRAY
```
Why type of `fixed_length` is `signed short`?
Somewhere in the code are negative values used?

vglavnyy commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -393,6 +393,84 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
+    return const_cast<mutable_return_type>(IndirectHelper<T>::Read(Data(), i));
+  }
+
+  // The raw data in little endian format. Use with care.
```
What does `little endian` mean?

mikkelfj replied:
Little endian is the most common storage order of integers with the least significant byte first.
Big endian has the opposite order. There are also other exotic orders found on old machines and in cryptography. The name originates from the childrens book Gulliver's travels where he visits the two neighbouring islands where one population eats their eggs with the big end up, and the other population eats their eggs with the little end up.

vglavnyy replied:
I read Gulliver :)
Lets I have BE machine and declare an array:
```cpp
int arr[3]={1,2,3};
auto pu8 = reinterpret_cast<const uint8_t*>(&arr[0]);
```
By nature, the `pu8` is neither BE nor LE.
It points to memory with BE integers (not to LE).

vglavnyy replied:
I was wrong, but have a new question.
Scalar is always represented in little-endian format. Array of scalars also `LE`.
```cpp
// The raw data in little endian format. Use with care.
uint8_t *Data() { return reinterpret_cast<uint8_t *>(&length_ + 1); }
const T *data() const { return reinterpret_cast<const T *>(Data()); }
```
How to use `data()` on `BE` machine?

mikkelfj replied:
I'll hazard a guess. This is a low-level pointer access method that is not endian safe. Hence the "use with care" comment. In flatcc for C, the method is called _get_ptr() where the safe method is _get(i) returning the field value.

For 8-bit elements it doesn't matter, but uint16_t would be unsafe. I have seen users getting into such problems on big endian hardware.

svenk177 replied:
Right you have to use `Get(i)` and `Mutate(i, val)`.

vglavnyy commented on src/idl_gen_cpp.cpp:
```diff
@@ -679,6 +679,10 @@ class CppGenerator : public BaseGenerator {
     if (IsScalar(type.base_type)) {
       return GenTypeBasic(type, user_facing_type) + afterbasic;
+    } else if (IsArray(type)) {
+      return beforeptr
```
code not formatted

vglavnyy commented on src/idl_gen_cpp.cpp:
```diff
@@ -2703,7 +2707,13 @@ class CppGenerator : public BaseGenerator {
+                         NumToString<short>(field.value.type.fixed_length) +
+                         "]"
+                       : "");
+      code_ += ("  {{FIELD_TYPE}}{{FIELD_NAME}}_{{ARRAY}};");
```
The code is not formatted.
Use `clang-format off/on` directives if a non-standard code formatting should be preserved.

vglavnyy commented on src/idl_gen_cpp.cpp:
```diff
@@ -2728,7 +2738,8 @@ class CppGenerator : public BaseGenerator {
     code_ += "  {{STRUCT_NAME}}() {";
-    code_ += "    memset(static_cast<void *>(this), 0, sizeof({{STRUCT_NAME}}));";
+    code_ +=
+      "    memset(static_cast<void *>(this), 0, sizeof({{STRUCT_NAME}}));";
```
code formatting (indent)

vglavnyy commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -393,6 +393,84 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
+  typedef VectorReverseIterator<const_iterator> const_reverse_iterator;
+
+  uoffset_t size() const { return length; }
+
```
The `uoffset_t size() const { return length; }` should be constexpr.

svenk177 replied:
Added constexpr.

mikkelfj commented:
Support added in FlatCC

https://github.com/dvidelabs/flatcc/commit/6ef1459ebc302f77bcd60f9e07989a116965bc49

binary format updated, notably removed supported for table fields with fixed length vectors and now call the feature Fixed Size Arrays.

https://github.com/dvidelabs/flatcc/blob/master/doc/binary-format.md

svenk177 commented:
I've made some improvements, especially support for struct arrays was introduced - though the C#, Java and Python ports are not there yet. Also char arrays are not implemented yet (any advice on how to integrate them?). 

I kept the `IndirectHelper` within the `Array` class as this allows reusing some existing code, e.g. see `PrintContainer`.

I also kept the stack of `Value`s as I'm a bit wary of alignment issues now that struct arrays have been introduced.

The length check is now capped at `0xFFFF` as the size has been changed to an `uint16_t`.

Hex values are now supported in the schema to match the behaviour in flatcc.

vglavnyy commented on src/idl_parser.cpp:
```diff
@@ -598,22 +598,20 @@ CheckedError Parser::ParseType(Type &type) {
       }
-      int64_t fixed_length = StringToInt(attribute_.c_str());
-      if (fixed_length < 1 || fixed_length > 0x7fff) {
+      int64_t fixed_length = 0;
```
`int64_t fixed_length = 0;` - uint16_t ?

svenk177 replied:
The return value of `StringToNumber` is an `int64_t`. The check prevents an over-/underflow.

vglavnyy replied:
https://github.com/google/flatbuffers/blob/bff7ffbc5130cd46caf33b76b4bb0593fcd15066/include/flatbuffers/util.h#L325

svenk177 replied:
D'oh missed it! Fixed.

vglavnyy commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -394,19 +394,15 @@ template<typename T> static inline size_t VectorLength(const Vector<T> *v) {
   typedef typename IndirectHelper<T>::return_type return_type;
-  typedef typename IndirectHelper<T>::mutable_return_type mutable_return_type;
+
+  constexpr uint16_t size() const { return length; }
```
FLATBUFFERS_CONSTEXPR

svenk177 replied:
Fixed.

mikkelfj commented:
Sounds good. FlatCC will be adding struct arrays and arrays of enums.

I think we need to discuss char arrays more before implementing them.

I think flatcc has a bug in that the C source cannot compile empty structs that are members of other structs. I'm not sure I'd bother fix it since it is rather messy. C does not support empty structs or other zero size types.

mikkelfj commented:
FlatCC has now support for scalars, enums and fixed size arrays of structs, i.e. arrays of all ordinary struct member types, including JSON parsing and printing. Char arrays are not supported.

https://github.com/dvidelabs/flatcc/tree/cf3f22eaf8082bccfd63efa8aeda41460b5be176

svenk177 commented:
I have added support for:

- enum arrays
- code generation for Java, C# with samples
- code generation for Python (not tested yet)

aardappel commented:
Sorry, what is the issue with char arrays?

`flatc` has an error for `"size 0 structs not allowed"`, I'd suggest `flatcc` do the same. So empty structs should not be something you have to worry about.

@rw for Python.

mikkelfj commented:
Empty structs are very useful for union types. But it would simplify a lot to remove them.

The issue with char arrays is:

- there is currently no char type
- if you use x:[string:15] that would imply 15 offsets to string objects.
- do you automatically insert a trailing zero, so x:[char:15] has size 16?
- do you have to verify that - currently structs are much simpler to verify.
- what about utf-8 code points not fitting exactly - same with strings, but more prominent with char arrays where you have a forced limit and truncation is then non-trivial.

Char arrays could also be just chars with no implied zero termination, or it could be optional and part of the specified size.

aardappel commented:
@mikkelfj well, `flatc` has already removed them :)

Ok, I guess we should be clear what we mean with `char` then, as to me that's the same a FlatBuffer's `byte`. FlatBuffers has no concept of `char` in the unicode sense.

Also not sure what you mean by `[string:15]`, as this PR only applies to scalars / fixed size things. Not understanding the rest of it either. Did I miss a discussion somewhere? :)

mikkelfj commented:
[string:15] has been suggested earlier, by you I think. I suggested char at that time (some chat a while ago AFAIR). My of raising it here was merely to agree on not using that syntax.

I'm not arguing against byte char, we just need to define what it means and how to handle edge cases, and notably 0 termination, if any.

aardappel commented:
If the idea is to allow `[byte:16]` also to contain 0-terminated strings, I'd suggest it be natural to make the terminator included in the size, so max 15 ascii characters (and who knows how many utf-8?) can be stored.

Does this need explicit support though? Couldn't this just be a helper function?

And in the C++ world, with `string_view` fast becoming the favored way to pass strings around, null-termination is unnecessary.

mikkelfj commented:
I wasn't aware of `string_view` me being mostly in C. But it is interesting because I have independently moved almost entirely to use string tokens which is my own <str, size> struct that I pass around (flatcc being an early example that has gone all the way to use token types), and it is much much simpler and more efficient. I do not assume any terminator within or after the token because that would prevent you from pointing into an existing string or text buffer.

So from that perspective I'm fine with just making it equivalent to [byte:16] and let the application interpret the data with not termination guarantees. The `char` type would be a hint to code generation to use string types. The buffer might contain separate length information such as a single byte length.

The verifier would not consequently not verify any content including presence or absence of any <32 ACSII control codes.

mikkelfj commented:
If we introduce char, is it then only in [char:n] form, or could structs have char fields? I'd rather only have the array form.

If `char` is to be considered as a scalar, is it byte or ubyte?
I'd prefer ubyte (= uint8).

svenk177 commented:
- Fixed an issue in Python code generation and added a sample
- Added a proposal for string handling - you can use any type and reinterpret it as a string, termination is enforced by default, but may be disabled via a template parameter

mikkelfj commented:
> Added a proposal for string handling 

You at least need a dedicated [char:<n>] syntax for parsing and printing JSON.

Adding a default expectation of 0 termination is dangerous if it is not part of the verification process, unless the operation has an internal check.

svenk177 commented:
> Adding a default expectation of 0 termination is dangerous if it is not part of the verification process, unless the operation has an internal check.

There is no expectation (unless you turn it off) - the `GetAsString` method force-writes a null terminator into the buffer.

mikkelfj commented:
I'm adding [char:n] to flatcc and expose it as a char array in C. The flatcc parser will fail any use of char that is not of the form [char:n]. This avoids discussions about char fields in tables and structs, and vectors of chars where ':' is absent. I also avoids default initializers of char type.

It will behave largely like [byte:n], (signed type because C char is signed).

JSON parsing will by default allow accept shorter strings and zero pad. It will also accept exact matches without adding zero-termination. Truncation will result in an error unless specified by runtime flag. If truncation is enabled, it might result in invalid UTF-8. JSON parsing will accept embedded null characters if properly escaped with `\x00`.

JSON printing will print up to and excluding the first null character or the full size, whichever comes first, but will print other control characters in their escaped form.

Note that truncation of chars is different from other arrays: a [byte:10] array will expect 10 elements unless a runtime flag allows for shorter arrays. I wonder if this should be changed such that shorter arrays are always permitted? This would avoid a special case and probably be more useful. In that case runtime flags would be: disallow_array_underflow, skip_array_overflow., or something like that.

mikkelfj commented:
Actually, JSON printing should strip trailing zeroes and print the rest. This will capture embedded null characters.

svenk177 commented:
> I'm adding [char:n] to flatcc and expose it as a char array in C.

I've added the [char:n] syntax. A new BASE_TYPE_CHAR_ARRAY was introduced to distinguish it from byte. The C++ generator uses `strncpy` instead of `Mutate` to initialize a character array.

> ..., (signed type because C char is signed).

This is actually not true.

> JSON printing will print up to and excluding the first null character or the full size, whichever comes first, but will print other control characters in their escaped form.

I would prefer to print the whole array including the null-bytes for symmetry.

mikkelfj commented:
> I would prefer to print the whole array including the null-bytes for symmetry.

That doesn't work well for shorter strings. But the end result has symmetry:
```
struct Foo {
    bar:[char:5];
}
```
JSON parser examples:
```
{ "bar": "R2D2" }
{ "bar": "XYZZY" }
{ "bar": "K2" }
{ "bar": "K2\x00A1" }
```
These are equivalent to parsing
```
{ "bar": "R2D2\x00" }
{ "bar": "XYZZY" }
{ "bar": "K2\x00\x00\x00" }
{ "bar": "K2\x00A1\x00\x00" }
```
and printing with the rule of stripping trailing zeros but not embedded zeroes:
```
{ "bar": "R2D2" }
{ "bar": "XYZZY" }
{ "bar": "K2" }
{ "bar": "K2\x00A1" }
```
which is the same as the first example parsed.

For other array types I would pad with zeroes but print all zeroes. I just doesn't seem very user friendly with strings. Especially if the buffer is fairly large, like `surname: [char:120]`.

mikkelfj commented:
> > ..., (signed type because C char is signed).
> 
> This is actually not true.

It depends on the compiler.

mikkelfj commented:
I'm a bit concerned about mutating the buffer for string access - is there a cast to mutable buffers or other means of making the user aware of this. Note that mutating a buffer could result in crash if the buffer is constructed with overlapping data types. For example, overlapping with a vtable offset.

mikkelfj commented:
> I would prefer to print the whole array including the null-bytes for symmetry.

If you do that, FlatCC would accept the null characters, so it will not break anything. But I suggest rethinking this based on the above.

svenk177 commented:
> > I would prefer to print the whole array including the null-bytes for symmetry.
> 
> If you do that, FlatCC would accept the null characters, so it will not break anything. But I suggest rethinking this based on the above.

`"s":` "Hel\x00lo"` this would break after parsing -> printing -> parsing.
I think the proper solution would be to print all embedded null characters, but skip all null characters at the end.

svenk177 commented:
> I'm a bit concerned about mutating the buffer for string access - is there a cast to mutable buffers or other means of making the user aware of this. Note that mutating a buffer could result in crash if the buffer is constructed with overlapping data types. For example, overlapping with a vtable offset.

Is that really something that is allowed?

mikkelfj commented:
> Is that really something that is allowed?
No, but hackers are rarely concerned about what is permitted. The thing is that the verifier cannot detect it.

mikkelfj commented:
> > "s": "Hel\x00lo"` this would break after parsing -> printing -> parsing.
> I think the proper solution would be to print all embedded null characters, but skip all null characters at the end.

Not it won't. Because the printing rule is to strip trailing zeroes, not embedded zeroes. That is, strlen cannot be used for this. Since the length is fixed, it is easy to do a reverse scan.
```
char buf[10] = { 0 };
size_t n = sizeof(buf);
memcpy(buf, "Hel\x0lo", 6);
while (n > 0 && buf[n - 1] != '\0') --n;
printf("\"%.*s\"", (int)n, buf);
```

svenk177 commented:
> > > "s": "Hel\x00lo"` this would break after parsing -> printing -> parsing.
> > > I think the proper solution would be to print all embedded null characters, but skip all null characters at the end.
> 
> Not it won't. Because the printing rule is to strip trailing zeroes, not embedded zeroes. 

This is exactly what I said. Your prior comment however contradicts that:

>JSON printing will print up to and excluding the first null character or the full size, whichever comes first...

> The thing is that the verifier cannot detect it.

Wouldn't any mutable buffer be susceptible to it then? Why can't the verifier detect that something is overlapping?

mikkelfj commented:
> This is exactly what I said. Your prior comment however contradicts that:

If we agree now, there is no point in disagreeing :) But my first statement that you quoted was followed by another where I realise that that was a bad idea.

> Wouldn't any mutable buffer be susceptible to it then? 

Yes, and this is why FlatCC is not supporting this, although it does provide buffer pointers and endian conversion functions if you really must. Mutation is rather limited anyway.

When you mutate, you should have inside knowledge of your buffer. FlatCC supports sorting existing buffers, which is also dangerous, but it is safe if you just created the buffer yourself.

> Why can't the verifier detect that something is overlapping?
It could, but it would be an order of magnitude more expensive and require external memory to remember memory areas that have been checked.

As it is, it is safe to access buffers with overlap as long as you do not mutate. Almost safe - you can still enter exponentially many tree traversals.

svenk177 commented:
> > This is exactly what I said. Your prior comment however contradicts that:
> 
> If we agree now, there is no point in disagreeing :) But my first statement that you quoted was followed by another where I realise that that was a bad idea.

Great! I used your little snippet for inspiration :)

> 
> > Wouldn't any mutable buffer be susceptible to it then?
> 
> Yes, and this is why FlatCC is not supporting this, although it does provide buffer pointers and endian conversion functions if you really must. Mutation is rather limited anyway.

I see, so I guess we rather need a `check()` function for `c_str()`. `str()` is safe as it copies the array into a `std::string`.

mikkelfj commented:
> while (length > 0 && s[length - 1]) length--;

meanwhile I fixed my snippet, you should too:

` while (length > 0 && !s[length - 1]) length--;`

svenk177 commented:
> ` while (length > 0 && !s[length - 1]) length--;`

I just found that as well :)

mikkelfj commented:
FlatCC now supports fixed size char arrays and JSON parser will now by default accept truncated arrays but for char arrays and other array types.

https://github.com/dvidelabs/flatcc/commit/6d5690da1d8cd8743b7c883b6e76b9913499a627

mikkelfj commented:
@aardappel I have removed support for empty structs in flatcc.

svenk177 commented:
I've added a `check()` function to tell if calling `c_str()` is safe or not. The ForceTermination "feature" has been removed.

mikkelfj commented:
Just a suggestion: perhaps a safe_c_str() that returns "" if there is no null terminator. Also, check could just check that last byte of the string since if there are are only embedded nulls, the result would be misleading.

In the flatcc binary doc I defined char arrays to require zero padding but not termination. This is compatible with strncpy behaviour. There shouldn't ever be trailing garbage.

mikkelfj commented:
I was also considering adding a length function to the C generated code, similar to one the JSON printer uses, but I didn't really know what to call it and chose just to document that users could do that if it suits them.

svenk177 commented:
> Just a suggestion: perhaps a safe_c_str() that returns "" if there is no null terminator. 

Will add!

> Also, check could just check that last byte of the string since if there are are only embedded nulls, the result would be misleading.

I thought about that too, but an embedded null-characted would at least make it safe to call c_str(). Hmm maybe provide two versions?

mikkelfj commented:
> Hmm maybe provide two versions?

I'm not really familiar with the C++ API. But there is the string_view option, and a raw pointer access, so I'm not sure one should pretend there is a c_str if it isn't really a c_str. The user can do low lever stuff as needed with raw pointer access. Note that unlike most other arrays, char arrays are endian safe and therefore safe to access directly. FlatCC documents that and has _get_ptr, as it has on all fields. In fact, FlatCC only has _get_ptr because there is no string class. Clearly less safe, but that is C for you.

svenk177 commented:
I've seen some templated classes out there which assume that a c_str() function exists, so I would prefer to keep it.

svenk177 commented:
Builds finally green on all targets!

vglavnyy commented:
I have cloned the repo for tests and have got errors (MSVC2017/19):

>\tests\monster_extra_generated.h(38,1): error C2220:  warning treated as error - no 'object' file generated
\tests\monster_extra_generated.h(38,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(89,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(91,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(150,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(153,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(155,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(240,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data
\tests\arrays_test_generated.h(242,1): warning C4267:  'argument': conversion from 'size_t' to 'flatbuffers::uoffset_t', possible loss of data

vglavnyy commented:
`reinterpret_cast<flatbuffers::Array<int32_t, 2> *>(a_)`:
1. Is UB if `a_` is not aligned properly.
2. Is type-punning => UB (common, not resolved problem)

svenk177 commented:
> I have cloned the repo for tests and have got errors (MSVC2017/19):

Thank you very much for testing! I will look into it tomorrow.

> 1. Is UB if a_ is not aligned properly.

To fix this we would need to fix EndianScalar, ReadScalar and WriteScalar to operate on `char` pointers.
Vectors are currently also prone to this.

> 2. Is type-punning => UB (common, not resolved problem)

`alignof(flatbuffers::Array<int32_t,1>)` equals to 1 as the underlying type is an uint8_t. uint8_t is defined as an unsigned char, which is exempted from the type-punning rule.
I don't know however if this is allowed:
`struct A { char c_; }; char c; A *a = reinterpret_cast<A*>(&c);`
If this is not an aliasing violation, then I think that we're safe here.
Maybe some language lawyer can clear this up :)

svenk177 commented:
- I changed the generated type to `uoffset_t`, @vglavnyy could you test again?

- I changed the internal type from a single `uint8_t` to an array, such that `sizeof(Array<int, 4>) == sizeof(int[4])`

mikkelfj commented:
The array should align to the element type, e.g uint32_t for flatbuffer int type.

mikkelfj commented:
int has no standard size in C/C++

svenk177 commented:
> int has no standard size in C/C++

Yeah, this was just examplary. You can replace `int` with `T` :)

vglavnyy commented:
@svenk177, the compilation has done without errors.
Maybe, split this PR into two parts (array and array-string)?
What do you think?

aardappel commented:
I would prefer not to introduce the concept of `char` or a char array. We can work with the existing `byte` type. Allowing the JSON parser to read zero padded strings into a `[byte]` does not need a new type.

mikkelfj commented:
That would require attributes for json parsing.

mikkelfj commented:
From a C perspective it is very convenient to have the char array in the struct as strncpy works out of the box.

svenk177 commented:
> Maybe, split this PR into two parts (array and array-string)?
> What do you think?

I have rebased and squashed it into three parts. Is that ok?

> I would prefer not to introduce the concept of char or a char array.

I have to agree with @mikkelfj that it fits nicely into C/C++ char array concept and also makes the intent explicit.

mikkelfj commented:
On the char equivalence to byte, I'd prefer to make it equivalent to ubyte since signed chars aren't very useful. This is what I documented in the binary doc:
https://github.com/dvidelabs/flatcc/blob/master/doc/binary-format.md

Old text on strings:
>Strings are stored as vectors of type ubyte_t, i.e. 8-bit elements

New text on char arrays:

>Fixed size arrays also support char arrays. The char type is similar to the ubyte or uint8 type but a char can only exist as a char array like x:[char:10]. Chars cannot exist as a standalone struct or table field, and also not as a vector element. Char arrays are like strings, but they contain no length information and no zero terminator. They are expected to be endian neutral and to contain ASCII or UTF-8 encoded text zero padded up the array size. Text can contain embedded nulls and other control characters. In JSON form the text is printed with embedded null characters but stripped from trailing null characters and a parser will padd the missing null characters.

aardappel commented:
@svenk177 3 commits doesn't do anything, as the result of this PR will be squashed anyway. What I said is to remove the `char` functionality from this PR and put it in a follow-up, as I am not convinced it is necessary.

@mikkelfj C++ has char as signed by default on all common systems, so using unsigned would create all sorts of clashes. Either way, we can discuss the need for this (or not) separate from this PR.

mikkelfj commented:
I agree that byte vs ubyte doesn't materially impact this PR. The unsignedness only matters if you choose to support a JSON numeric array or if you introduce a char field with a default value with a default value > 127 in the schema and permit such a construct in the first place.. For C I cast it to char regardless, so the clash you mention can be ignored - but perhaps that should be clarified in the text.

svenk177 commented:
@aardappel I have moved char array support into PR #5354.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -401,35 +401,153 @@ static inline bool StringLessThan(const char *a_data, uoffset_t a_size,
-struct String : public Vector<char> {
-  const char *c_str() const { return reinterpret_cast<const char *>(Data()); }
-  std::string str() const { return std::string(c_str(), size()); }
+template<typename T> struct StringBase : public T {
```
This change I believe was only needed to support `char` related features. Please keeps `String` as it was.

svenk177 replied:
This may also be used to store a string in a byte array, but I can remove it if you wish.

aardappel replied:
For now, please do. I'd like to see the impact of such changes separately.

svenk177 replied:
Removed.

aardappel commented on src/idl_parser.cpp:
```diff
@@ -1234,6 +1279,58 @@ CheckedError Parser::ParseVector(const Type &type, uoffset_t *ovalue,
+
+  if (length != count) return Error("Fixed-length array size is incorrect.");
+
+  for (uint16_t i = 0; i < count; i++) {
```
There's no good reason to hardcode `uint16_t` in places like this, this could just be a `size_t` or `uoffset_t`, which will lead to less headaches down the line if we ever change these limits. Either that, or make a typedef for `uint16_t` that clearly indicates we're talking about array sizes here, but that's probably overkill here too.

svenk177 replied:
Changed to `size_t`.

rw commented:
@aardappel @svenk177 I don't see tests for the Python code. Can you remove the Python changes from this PR, and submit another PR when you have time to write tests?

svenk177 commented:
@rw, there are tests (or at least samples with asserts) in `sample/python_array_sample.sh, sample/java_array_sample, sample/csharp_array_sample.sh`, though I have to admit that they are not integrated into CI.

aardappel commented:
@rw @svenk177 I agree there should be tests, and the `samples/` dir is not for tests. It be better to move whatever is there to the `tests` dir, so it can be run as part of the other tests, and part of the existing language source files and scripts, to keep it simple. Given that this is a new feature that is not fully supported, there is no need to have a sample for it.

Random idea: for languages that do not support this feature yet, would it make sense to have the schema parser insert extra fields in the `StructDef`? e.g. `myarray[int:2]` causes `myarray0:int` and `myarray1:int` to be added, which means the generators for those languages will generate working code without changes. Downside of course is that now implementing proper support becomes a breaking change. Maybe not a good idea?

svenk177 commented:
Added tests for Python, Java and C#.

>Random idea: ...

Generally this sounds nice, but I fear its usefulness is somewhat limited, e.g. `arr:[int:512]` -> you'd have to pass 512+ arguments to a function and AFAIK Java doesn't even allow this.

aardappel commented:
Yes, it was just a random idea, maybe leave it for now. Better actual support gets implemented.

Can you remove your tests from `samples` ? They only need to be in `tests`. I don't think we need samples for this feature yet, as I mentioned above.

aardappel commented:
There's still a schema in `samples/` but otherwise looks ready to merge to me.

Anyone else have final comments? @rw?

aardappel commented:
I see no further comments, so this is ready to merge, though it looks like it needs to be rebased first.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -378,7 +382,8 @@ class GeneralGenerator : public BaseGenerator {
   // Casts necessary to correctly read serialized data
   std::string DestinationCast(const Type &type) const {
-    if (type.base_type == BASE_TYPE_VECTOR) {
+    if (type.base_type == BASE_TYPE_VECTOR ||
```
Optional:
`IsArray(type) || IsVector(type)` - ?
Or-short: `IsSeries(type)`

Here and whenever direct comparisons with `BASE_TYPE_VECTOR` and/or `BASE_TYPE_ARRAY` are used.

svenk177 replied:
Introduced `IsSeries` and `IsVector`, though I did not change existing comparisons against `BASE_TYPE_VECTOR`.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -650,20 +657,42 @@ class GeneralGenerator : public BaseGenerator {
+        array_cnt++;
+      }
+
+      if (IsStruct(type)) {
```
Maybe, for readability, introduce new local constants?
```cpp
      const auto field_type = field.value.type;
      const auto array_field = IsArray(field_type); 
      const auto type = array_field ? field_type.VectorType() : field_type;
      const auto array_cnt = array_field ? (array_count + 1) : array_count;
```
Please, don't use reassignment if it is possible (`type` variable).

svenk177 replied:
Changed.

vglavnyy commented on src/idl_gen_text.cpp:
```diff
@@ -69,26 +69,27 @@ bool Print(T val, Type type, int /*indent*/, Type * /*union_type*/,
   text += "[";
   text += NewLine(opts);
-  for (uoffset_t i = 0; i < v.size(); i++) {
+  for (uoffset_t i = 0; i < size; i++) {
```
Why not use iterators here (iterate Container `c`)?
Is pointer arithmetic really needed?

svenk177 replied:
I reused the existing code here. I tried iterators here, unfortunately the templated `Print` function doesn't mix well with iterators. The `if(IsStruct)` statement is still required to do the required `reinterpret_cast` for structs. Another issue is that the size of an array can not be determined statically, hence `end` / `cend` is not usable here.

vglavnyy commented on src/idl_parser.cpp:
```diff
@@ -1234,6 +1279,58 @@ CheckedError Parser::ParseVector(const Type &type, uoffset_t *ovalue,
+
+  if (length != count) return Error("Fixed-length array size is incorrect.");
+
+  for (size_t i = 0; i < count; i++) {
```
Is it possible to use reverse iterator over `stack`?
The reverse iterator will remove `stack.pop_back()`.

svenk177 replied:
Using iterators now.

vglavnyy commented on tests/test.cpp:
```diff
@@ -2671,6 +2681,47 @@ void CreateSharedStringTest() {
+              reinterpret_cast<const void *>(&nStruct1), sizeof(nStruct1));
+  std::memcpy(reinterpret_cast<void *>(&nStruct[1]),
+              reinterpret_cast<const void *>(&nStruct2), sizeof(nStruct2));
+
```
`std::vector<MyGame::Example::NestedStruct> nStruct` - ?
With `aStruct(2, array, 12, &nStruct[0]);`

svenk177 replied:
Changed to std::vector.

vglavnyy commented on src/idl_gen_python.cpp:
```diff
@@ -402,21 +428,43 @@ class PythonGenerator : public BaseGenerator {
+          code += IsArray(field.value.type) ? "    " : "";
+          code += indent + "    builder.Prepend" + GenMethod(field) + "(";
+          code += nameprefix + MakeCamel(NormalizedName(field), false);
+          for (size_t i = 0; in_array && i < index + IsArray(field.value.type) ? 1 : 0; i++) {
```
Please, set priority in the condition checking with `()` brackets.
Or remove `IsArray(field.value.type) ? 1 : 0` out of loop.

PVS-Studio high-pri warning:
> V502 Perhaps the '?:' operator works in a different way than it was expected. The '?:' operator has a lower priority than the '+' operator. idl_gen_python.cpp 463

https://www.viva64.com/en/w/v502/print/

svenk177 replied:
Good catch!

mikkelfj replied:
You can use `!!IsArray(field.value.type)` instead of `(IsArray(field.value.type) ? 1 : 0)`

vglavnyy commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,589 @@
+  TestStruct() {
+    memset(static_cast<void *>(this), 0, sizeof(TestStruct));
+  }
+  TestStruct(uint8_t _a, const uint16_t *_b, const uint32_t *_c, const double *_d, uint16_t _e, const NestedStruct &_f)
```
PVS-Studio:
>V730 Not all members of a class are initialized inside the constructor. Consider inspecting: b_, c_, d_. arrays_test_generated.h 212

False positive?
Generated code never called in tests (I don't found calls).

svenk177 replied:
Removed `TestStruct` from the `fbs`.

vglavnyy commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,589 @@
+  NestedStruct() {
+    memset(static_cast<void *>(this), 0, sizeof(NestedStruct));
+  }
+  NestedStruct(const int32_t *_a, TestEnum _b, const TestEnum *_c)
```
PVS-Studio:
>V730 Not all members of a class are initialized inside the constructor. Consider inspecting: a_, c_. arrays_test_generated.h 85

`NestedStruct` never called in cpp code.

svenk177 replied:
You mean the default constructor is not called? I think it's fair to assume that it works :)

vglavnyy replied:
No, the second constructor :)
Guess I missed it in the test on the first review.

vglavnyy commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,589 @@
+  ArrayStruct() {
+    memset(static_cast<void *>(this), 0, sizeof(ArrayStruct));
+  }
+  ArrayStruct(float _a, const int32_t *_b, int8_t _c, const NestedStruct *_d)
```
PVS-Studio:
>V730 Not all members of a class are initialized inside the constructor. Consider inspecting: b_. arrays_test_generated.h 142

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -1012,6 +1074,7 @@ class GeneralGenerator : public BaseGenerator {
             code += offset_prefix + getter + "(o + " + lang_.accessor_prefix;
             code += "bb_pos) : null";
             break;
+          case BASE_TYPE_ARRAY:
```
```cpp
case BASE_TYPE_ARRAY: FLATBUFFERS_FALLTHROUGH(); // fall thru
case BASE_TYPE_VECTOR: {
```

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -1221,21 +1294,22 @@ class GeneralGenerator : public BaseGenerator {
         if (IsScalar(field.value.type.base_type) ||
-            (field.value.type.base_type == BASE_TYPE_VECTOR &&
-             IsScalar(field.value.type.VectorType().base_type))) {
+            (is_series && IsScalar(field.value.type.VectorType().base_type))) {
```
`IsScalar(underlying_type)` ? (guess)

svenk177 replied:
Indeed!

vglavnyy commented on src/idl_gen_python.cpp:
```diff
@@ -224,6 +224,29 @@ class PythonGenerator : public BaseGenerator {
+      code += NumToString(field.value.type.fixed_length) + ")]\n";
+    }
+  }
+
```
`field.value.type.VectorType()` four times refered. Add `vec_type = field.value.type.VectorType()` ?

svenk177 replied:
Changed.

vglavnyy commented:
I'm sorry, `fall thou` was not necessary. I have checked it with gcc 8.3 and clang 7.1.

svenk177 commented:
> 
> 
> I'm sorry, `fall thou` was not necessary. I have checked it with gcc 8.3 and clang 7.1.

We might as well keep it to make the intent clear :)

vglavnyy commented on src/idl_gen_text.cpp:
```diff
@@ -144,6 +159,27 @@ bool Print<const void *>(const void *val, Type type, int indent,
         // clang-format on
       }
       break;
+    case BASE_TYPE_ARRAY: {
```
Above, BASE_TYPE_VECTOR redefine `type` to `type = type.VectorType();` and use it as an element type.
Better rewrite BASE_TYPE_ARRAY branch in the uniform style.
Maybe introduce `elem_type` in both branches.

vglavnyy commented on src/idl_parser.cpp:
```diff
@@ -1234,6 +1277,57 @@ CheckedError Parser::ParseVector(const Type &type, uoffset_t *ovalue,
+  for (auto it = stack.rbegin(); it != stack.rend(); it++) {
+    auto &val = *it;
+    switch (val.type.base_type) {
+// clang-format off
```
Move `// clang-format on` before `switch` and `off` after `}' of the switch.
Then the position of `// clang-format on`  will be stable.

vglavnyy commented:
Squash, rebase to master and merge?

svenk177 commented:
Squashed and rebased!

aardappel commented:
@vglavnyy You think its good to go in?

vglavnyy commented:
@svenk177 
Can you extend the array's test with json import/export?
1. Load data from json file, GetArrayTable, VerifyArrayTableBuffer, generate json2.
2. Parse json2 from string, GetArrayTable, VerifyArrayTableBuffer, compare with table1.

You can exam the Monster  (AccessFlatBufferTest or ParseAndGenerateTextTest).
I have checked import and export using flatc: [tests.zip](https://github.com/google/flatbuffers/files/3268754/tests.zip)
Works well, but it is better to add to the test.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -681,29 +703,67 @@ class GeneralGenerator : public BaseGenerator {
     for (auto it = struct_def.fields.vec.rbegin();
          it != struct_def.fields.vec.rend(); ++it) {
       auto &field = **it;
+      const auto field_type = field.value.type;
```
`const auto &field_type` - ?

svenk177 replied:
Added.

vglavnyy commented on src/idl_gen_python.cpp:
```diff
@@ -380,12 +404,17 @@ class PythonGenerator : public BaseGenerator {
         it != struct_def.fields.vec.end(); ++it) {
       auto &field = **it;
-      if (IsStruct(field.value.type)) {
+      const auto field_type = field.value.type;
```
const ref?

svenk177 replied:
Added.

vglavnyy commented on src/idl_gen_python.cpp:
```diff
@@ -402,22 +431,50 @@ class PythonGenerator : public BaseGenerator {
     for (auto it = struct_def.fields.vec.rbegin();
         it != struct_def.fields.vec.rend(); ++it) {
       auto &field = **it;
+      const auto field_type = field.value.type;
```
const ref-?

svenk177 replied:
Added.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -656,20 +661,37 @@ class GeneralGenerator : public BaseGenerator {
          it != struct_def.fields.vec.end(); ++it) {
       auto &field = **it;
-      if (IsStruct(field.value.type)) {
+      const auto field_type = field.value.type;
```
cont ref - ?

svenk177 replied:
Added.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -656,20 +661,37 @@ class GeneralGenerator : public BaseGenerator {
-      if (IsStruct(field.value.type)) {
+      const auto field_type = field.value.type;
+      const auto array_field = IsArray(field_type);
+      const auto type = array_field ? field_type.VectorType()
```
`const auto &type = field_type.VectorType()` is valid C++ construction.
The lifetime of returned temporary will be prolonged as long as it's a const reference.
[GotW 88](https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/)

Better to use `const auto &` for VectorType() or remove it everywhere (idl_parser.cpp:1283, for example).
MSVC detects violation if `const` removed and generate a warning "A non-const reference may only be bound to an lvalue".

svenk177 replied:
Added.

vglavnyy commented on tests/monster_extra.fbs:
```diff
@@ -1,5 +1,9 @@
+struct ArrayStruct {
+    array:[int:2];
+}
+
```
We can remove the array from MosterExtra.
It was added before arrays_test.fbs was introduced.

svenk177 replied:
Removed.

vglavnyy commented on tests/arrays_test_generated.h:
```diff
@@ -0,0 +1,411 @@
+  ArrayStruct() {
+    memset(static_cast<void *>(this), 0, sizeof(ArrayStruct));
+  }
+  ArrayStruct(float _a, const int32_t *_b, int8_t _c, const NestedStruct *_d)
```
This constructor and `NestedStruct(const int32_t *_a, TestEnum _b, const TestEnum *_c)` look extrimly dangerous.
Is it really needed in the generated code?
Can a developer write own builder method?

svenk177 replied:
I am aware, unfortunately C++ does not support the C99 syntax for required array lengths, i.e. 
`void func(int arr[static 15])`. Do you think we should use `std::array` here?
Own builder methods are not supported, but AFAIK they don't exist on `struct`s either.

mikkelfj replied:
Why are structs not exposed as basic C structs in addition to accessor methods? Then you can trivially get the size of any array.

vglavnyy commented:
The current implementation has a drawback: it is impossible to get the size (length) of a fixed array at compile time.
For example, how to get the size of `ArrayStruct.b` field?
I can deduce the type of `b`: `using ab_type = decltype(ArrayStruct().b())` - returns const pointer.
But I can't get access to `size()` without tricks.
The `Array` should be like `std::array<>` or generated code should declare type and size of fixed array fields.

@ svenk177 do you have an example of how to extract size at compile time?

svenk177 commented:
> @ svenk177 do you have an example of how to extract size at compile time?

`sizeof( decltype(*ArrayStruct().b()) ) / sizeof( decltype(*ArrayStruct().b()->data()) )`
works, but I have to admit that it is pretty clunky.

I don't think it is possible to make it behave like `std:array` as the function returns a `const` pointer, as you said.
Maybe introduce a template like: `ArraySize<decltype(ArrayStruct().b())>::size` ?

svenk177 commented:
I've found that the equality operators are broken (even without this patch, but my array comparison is faulty too)!

Look at `inline bool operator==(const MonsterT &lhs, const MonsterT &rhs)`.
It compares `(lhs.pos == rhs.pos)`. It turns out `pos` is a `flatbuffers::unique_ptr`, therefore this compares pointers not the value itself!

mikkelfj commented:
I didn't think of bfbs in flatcc. I can't seem to find any commits changing the binary flatbuffers schema? Are they in a different PR?

mikkelfj commented:
Never mind, I found the changes in 
https://github.com/google/flatbuffers/pull/5313/commits/4cae6c6074160cc207d7dd12ce524cc6094d38cf#diff-db35d829e5e236af29f9a061c8352dcb

mikkelfj commented on reflection/reflection.fbs:
```diff
@@ -32,6 +33,7 @@ table Type {
                      // If base_type == Union, UnionType, or integral derived
                      // from an enum, index into "enums" below.
+    fixed_length:uint16 = 0;   // Only if base_type == Array.
 }
```
> element:BaseType = None;  // Only if base_type == Vector.
-->
> element:BaseType = None;  // Only if base_type == Vector or base_type == Array.

svenk177 replied:
Added.

vglavnyy commented:
`template<typename T, uint16_t length, typename U = T> class Array`
What is type `U` for? To store an enum without explicit type declaration?
All modern compilers accept a declaration of the underlying type.
The Array is a new feature and we can declare new requirements.
This type `U` complicate design and maintenance.

vglavnyy commented:
This is proposal for compile-time reflection (type, size).
[0001-Compile-time-reflection.zip](https://github.com/google/flatbuffers/files/3278330/0001-Compile-time-reflection.zip)

Second-type constructor issue:
Maybe remove the initialization of arrays inside the constructor (replace by zero-memset)?
A developer can own write external initialization code using std::array, std::vector, std::span or manual mutation.

svenk177 commented:
> What is type U for?

For this schema: `enum TestEnum : byte { }` the code generator will NOT generate `enum TestEnum : uint8_t { }`, but it will generate `enum TestEnum { }` and use a `uint8_t` for storage and appropriate `static_cast`s to guarantee type-safety. I.e. `sizeof(TestEnum) != sizeof(uint8_t)`. I guess this is for compatibility with old compilers. Therefore `T = TestEnum` and `U = uint8_t`.
I'm surprised the test did not fail after you have removed the `U` type.

> This is proposal for compile-time reflection (type, size).

This looks great, I will include it.

> Second-type constructor issue:

I'd rather keep it than not offering one at all. Maybe a templated constructor can help here?

vglavnyy commented:
You can use `parser_.opts.scoped_enums` for generated tests.
Optionally, you can check that this flag is active if see a fixed array of enums:
https://github.com/google/flatbuffers/blob/a807fa9567c7302b2c446efb3bc7ee79d1a462b4/src/idl_gen_cpp.cpp#L954-L961

I think that enums without an explicit underlying type should become a legacy.

svenk177 commented:
> I think that enums without an explicit underlying type should become a legacy.

I agree, but unless VS10 will be deprecated I think its fair to keep it. It can be removed easily at a later date.

The Android compiler seems to be missing `tuple`. Do you have a workaround at hand?

vglavnyy commented:
Android compiler doesn't have `<tuple>`.
`#include <tupple>` isn't necessary.
But this does not 'resolve the issue. This weak compiler will fail on the `std::enable_if` or something like this.
I vote to remove compile-time features for the `Array` instead of adding ugly workaround.
Let the `Array` be just a slightly different `Vector`.
We can add it later.

svenk177 commented:
> This weak compiler will fail on the std::enable_if or something like this.

You are right. Dropped the patches!

I think fixing the equality operators should be done in a seperate issue. Do you agree?

vglavnyy commented:
@svenk177  Yes, compare operator for `Table` should be fixed in a separate issue.

Recommendations (optional) for this PR:
1. Replace `template<typename T, uint16_t length, typename U = T>` to ``template<typename T, uint16_t length>.
Check or require that `parser.opts.scoped enums` flag is active to ignore MSVC2010.
In the tests use PPD to avoid MSVC2010.

2.  Remove/replace `ArrayStruct(float _a, const int32_t *_b, int8_t _c, const NestedStruct *_d)` by
`ArrayStruct(float _a, int8_t _c)`.
This kind of constructor absolutely unusable in production code.
If one modifies `fbs` schema with increasing an array length it may cause an `out-of-range`.
Move an initialization into test and sample code.

If is ready notify **aardappel** for final review.

svenk177 commented:
@vglavnyy Requested changes commited.
@aardappel Review?

vglavnyy commented on src/idl_gen_cpp.cpp:
```diff
@@ -307,6 +307,24 @@ class CppGenerator : public BaseGenerator {
+        }
+      }
+    }
+
```
Why do you need to traverse `parser_.structs_.vec`?
Is it possible to use lazy checking? For example after line 2745 when a field of a struct is generated.

vglavnyy commented on src/idl_gen_general.cpp:
```diff
@@ -1210,9 +1279,9 @@ class GeneralGenerator : public BaseGenerator {
-                                   : field.value.type;
+        auto is_series = (IsSeries(field.value.type));
+        auto underlying_type =
+            is_series ? field.value.type.VectorType() : field.value.type;
```
`const auto &underlying_type` - ?

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -2717,10 +2725,22 @@ class CppGenerator : public BaseGenerator {
+      // Check if enum arrays are used without specifying --scoped-enums
+      if (!parser_.opts.scoped_enums && IsArray(field_type) &&
+          IsEnum(field_type.VectorType())) {
+        printf("--scoped-enums must be enabled to use enum arrays\n");
```
Not happy with this error for multiple reasons. First it would be better if `--scoped-enums` wasn't required. And if that really isn't possible, I'd prefer to trigger this error in the parser, where are all error handling happens. The generators don't really have a way to do errors, and this print + return false is a bit of a hackey way to do it.

vglavnyy replied:
@aardappel 
The idea to ignore untyped enums was mine. Initially, `Array` template had the support of enums without underlying type specification. But this third template parameter complicates the template and complicate compile-time helpers for the `Array` (compile-time functions like `size()` or other `constexpr` features).
There are a few C++ compilers without underlying type specification. I assumed that `Array` is a new feature needn't support of legacy. One should use a modern compiler to use `Array`.
Of course, this should be documented.

I can **restore** removed code if I'm not right.

Generator vs parser:
The flag `--scoped-enums` needed for C++ code generator only.
Other languages are independent of this flag.
One can use `flatc` to convert from json to binary and versa. 
The `--scoped-enums` isn't necessary for these conversions.

Another way is to silently add an underlying type to generated enums used with Array.
For example, the underlying type silently added for each int64/uint64 enum due to MSVC restrictions.

mikkelfj replied:
I thought enums were always explicitly typed. But if they are not, they should IMO default to 32 bit like `int`. I actually find it a bit annoying that I have to specify a type for enums since often it really doesn't matter.

mikkelfj replied:
umm - I'm talking about flatbuffer schema types - is this about C++ enums being typed (I have given up keeping track of that language) ?
In C the enums are represented as typecast integer defines in order to associate a type, .e.g. `#define Red INT32_C(1)`
Originally the were constants like `const int32_t Red = 1;` but that did not work in switch statements.

svenk177 replied:
I didn't like the `printf` in there either, therfore I moved the check for scoped-enums into the parser along with a check if it generates code for C++.

@mikkelfj C++11 introduced the following syntax `enum A : uint8_t { }`, i.e. `sizeof(A) == sizeof(uint8_t)`, unfortunately old compilers do not support it, hence the requirement for `--scoped-enums`

mikkelfj replied:
OK, thanks

svenk177 replied:
The now removed third parameter in the `Array` template class was intended for this purpose, but it was indeed a bit messy as @vglavnyy pointed out.

mikkelfj replied:
Just accidentally found this http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2008.pdf
It appers C might get typed enum in C2x, not that it matters since it would not be portable.

aardappel replied:
@vglavnyy if it is needed to keep things simple, then so be it. I'd still like the actual error checking to be moved to the parser though, if possible.

svenk177 replied:
@aardappel I already moved it into the parser. I had to squash the history as rebasing was not fun otherwise.

aardappel replied:
Ah thanks!

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -2745,34 +2765,44 @@ class CppGenerator : public BaseGenerator {
+        first++;
+        continue;
+      }
+
```
This PR introduces a lot of blank lines.. prefer to not have them by default, only if a method is really long. And even then, comments are probably a better separator.

svenk177 replied:
Removed most blank lines and added some comments.

aardappel commented:
Ok, that means we're once again ready to merge! I see @vglavnyy's last comments have been addressed, so I guess we can go ahead.

Thanks everyone for helping get this feature in!

aardappel commented:
This PR accidentally broke because of a PR by @vglavnyy I merged just before this one: https://github.com/google/flatbuffers/pull/5381 (makes Table/Struct internal access).
CI here: https://travis-ci.org/google/flatbuffers/builds/546979717
A bit of a coincidence :)

vglavnyy commented:
Run of `generate_code.sh` with updated `flatc` should fix it. There are no conflicts in the code.

svenk177 commented:
Thank you all for getting this in :)

aardappel commented:
Updated generated code in https://github.com/google/flatbuffers/commit/123c7a48907c5080a41768066cbd48b2574cdc00

aardappel commented:
@svenk177 @vglavnyy 

We have CI errors related to this PR: https://ci.appveyor.com/project/gwvo/flatbuffers/builds/27162701/job/3edsmdethte5y7w9

Note sure why that is failing now all of a sudden?

Looking into the errors and the code, the problem appears to be that `Array::Mutate` is being called with a `T` which is a struct.. even though `Mutate` was written for scalars only.

Relevant code:

https://github.com/google/flatbuffers/blob/4b870aca98eca22c56faa3245ada7014752f7b2c/tests/test.cpp#L2808-L2839

Note e.g. `aStruct.mutable_d()->Mutate(0, nStruct0);` which calls `Mutate` with a struct, which eventually goes into `WriteScalar` which should never take structs.

Schema: https://github.com/google/flatbuffers/blob/master/tests/arrays_test.fbs

We should a) move the function `AssertScalarT()` to top level, and call it from `Mutate` or maybe even `WriteScalar` and b) fix these tests. c) we could add a `MutateStruct` if we want, since a struct is already endian-correct, so can be written to the array without calling `WriteScalar`.

Possibly remotely related: https://github.com/google/flatbuffers/pull/5440 which also had some alignment based MSVC errors.

vglavnyy commented:
The test should use `GetMutablePointer`
https://github.com/google/flatbuffers/blob/4b870aca98eca22c56faa3245ada7014752f7b2c/include/flatbuffers/flatbuffers.h#L441-L449

Another way (without suppression of always-const warning):
```cpp
  // Change elements if you have a non-const pointer to this object.
  void Mutate(uoffset_t i, const T &val) {
    if (flatbuffers::is_scalar<T>::value) {
      FLATBUFFERS_ASSERT(i < size());
      WriteScalar(data() + i, val);
    }
    else {
      *(GetMutablePointer(i)) = val;
    }
  }
```
Or we can use SFINAE but I prefer if-else (C++17 has constexpr-if).

**Second**:
> Possibly remotely related: #5440 which also had some alignment based MSVC errors.

Yes, we have the alignment problem (was detected by ASAN).
Need to research this.

aardappel commented:
@vglavnyy thanks, yes that is also a nice way to do it. Though of course this code is still going to do strange things if `T` is not a struct either, so maybe nice to address at some point :)

tsingson commented:
this feature is cool, i will port to Go.
